k2/csrc/fsa_utils.cu

```diff
@@ -1475,30 +1476,189 @@ Array1<FloatType> BackpropGetBackwardScores(
+  NVTX_RANGE(K2_FUNC);
+  ContextPtr c = fsas.Context();
+
+  // We will be adding to the elements of `forward_scores_deriv`.
+  // `forward_scores_deriv_in` was just the derivative w.r.t. the output of
+  // GetForwardScores(), but because GetForwardScores() is recursive,
+  // the derivatives for later states contribute to those of earlier ones.
+  Array1<FloatType> forward_scores_deriv(forward_scores_deriv_in.Clone());
+  FloatType *forward_scores_deriv_data = forward_scores_deriv.Data();
+  const int32_t *fsas_row_splits1_data = fsas.RowSplits(1).Data(),
+                *fsas_row_ids1_data = fsas.RowIds(1).Data(),
+                *fsas_row_ids2_data = fsas.RowIds(2).Data();
+  const FloatType *forward_scores_data = forward_scores.Data();
+  const Arc *arcs = fsas.values.Data();
+
+  K2_CHECK_EQ(fsas.NumAxes(), 3);
+  K2_CHECK_EQ(state_batches.NumAxes(), 3);
+  K2_CHECK_EQ(leaving_arc_batches.NumAxes(), 4);
+  int32_t num_fsas = fsas.Dim0(), num_states = fsas.TotSize(1),
+          num_arcs = fsas.TotSize(2);
+  Array1<FloatType> arc_scores_deriv(c, num_arcs);  // will return this.
+  FloatType *arc_scores_deriv_data = arc_scores_deriv.Data();
+
+  const FloatType negative_infinity =
+      -std::numeric_limits<FloatType>::infinity();
+  RaggedAxis0Splitter<int32_t> arc_batches_splitter(leaving_arc_batches);
+  int32_t num_batches = leaving_arc_batches.Dim0();
+
+  if (log_semiring) {
+    // For each batch of states, from end to start (opposite direction to
+    // GetForwardScores())...
+
+    for (int32_t b = num_batches - 1; b >= 0; --b) {
+      int32_t arc_begin;
+      Ragged<int32_t> leaving_arc_batch =
+          arc_batches_splitter.GetElement(b, &arc_begin);
+      int32_t *leaving_arc_batch_data = leaving_arc_batch.values.Data();
+      int32_t num_arcs = leaving_arc_batch.NumElements();
+
+      Ragged<FloatType> leaving_arc_deriv(leaving_arc_batch.shape);
+      FloatType *leaving_arc_deriv_data = leaving_arc_deriv.values.Data();
+
+      K2_EVAL(
+          c, num_arcs, lambda_set_arc_deriv_etc, (int32_t arc_idx) {
+            int32_t fsas_arc_idx012 = leaving_arc_batch_data[arc_idx];
             Arc arc = arcs[fsas_arc_idx012];
             int32_t dest_state_idx1 = arc.dest_state,
                     src_state_idx1 = arc.src_state,
                     src_state_idx01 = fsas_row_ids2_data[fsas_arc_idx012],
                     state_idx0x = src_state_idx01 - src_state_idx1,
                     dest_state_idx01 = state_idx0x + dest_state_idx1;
+            FloatType src_score = forward_scores_data[src_state_idx01],
+                      arc_end_score = src_score + arc.score,
+                      dest_score = forward_scores_data[dest_state_idx01];
+            // alpha = d(dest_score) / d(arc_end_score)
+            FloatType alpha = exp(arc_end_score - dest_score),
+                      arc_deriv =
+                          alpha * forward_scores_deriv_data[src_state_idx01];
```

Comment:
Should it be
```cpp
alpha * forward_scores_deriv_data[dest_state_idx01];
```
?
Correct me if I am wrong.

Reply:
yes you are right.  @qindazhu can you fix in next commit?

Reply:
I have fixed it locally